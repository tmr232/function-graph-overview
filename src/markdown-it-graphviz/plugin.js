import { Graphviz } from "@hpcc-js/wasm-graphviz";

export async function GraphvizDotPlugin() {
  const graphviz = await Graphviz.load();

  return (md) => {
    // Save the original fence renderer
    const defaultFence = md.renderer.rules.fence;

    // Override the fence renderer
    md.renderer.rules.fence = (tokens, idx, options, env, self) => {
      const token = tokens[idx];

      if (!token) {
        return "";
      }

      if (token.info.trim() === "dot") {
        const code = token.content.trim();

        const svg = graphviz.dot(code);

        return `<div class="dot-graph">
        ${svg}
      </div>`;
      }

      // For other languages, use the default renderer
      if (!defaultFence) {
        return "";
      }
      return defaultFence(tokens, idx, options, env, self);
    };
  };
}
