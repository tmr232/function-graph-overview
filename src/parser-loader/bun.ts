import { Parser, Language as TreeSitterLanguage } from "web-tree-sitter";
import type { Language } from "../control-flow/cfg.ts";
import { wasmMapping } from "./wasmMapping.ts";

export async function initializeParser(
  language: Language,
): Promise<{ parser: Parser; language: TreeSitterLanguage }> {
  await Parser.init();
  const parserLanguage = await TreeSitterLanguage.load(wasmMapping[language]);
  const parser = new Parser();
  parser.setLanguage(parserLanguage);
  return { parser, language: parserLanguage };
}
