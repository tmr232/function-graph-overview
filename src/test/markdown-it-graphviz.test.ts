import {test, describe, expect} from "bun:test";

import MarkdownIt from 'markdown-it'
import {GraphvizDotPlugin} from "../markdown-it-graphviz/plugin.ts"


describe("Render DOT Samples", () => {
  const md = new MarkdownIt()
    .use(GraphvizDotPlugin);

  test("Render DOT", ()=>{
    const result  = md.render(
`
Here's a DOT snippet!
\`\`\`dot
digraph G {

  subgraph cluster_0 {
    style=filled;
    color=lightgrey;
    node [style=filled,color=white];
    a0 -> a1 -> a2 -> a3;
    label = "process #1";
  }

  subgraph cluster_1 {
    node [style=filled];
    b0 -> b1 -> b2 -> b3;
    label = "process #2";
    color=blue
  }
  start -> a0;
  start -> b0;
  a1 -> b3;
  b2 -> a3;
  a3 -> a0;
  a3 -> end;
  b3 -> end;

  start [shape=Mdiamond];
  end [shape=Msquare];
}
\`\`\`
`
    );
    expect(result).toMatchSnapshot();
  });
})



describe("Render DOT-CFG Samples", () => {
  const md = new MarkdownIt()
    .use(GraphvizDotPlugin);

  test("Render CFG", ()=>{
    const result  = md.render(
      `
\`\`\`dot-cfg
entry [class="entry"]
return1 [class="exit"]
return2 [class="exit"]
return4_2 [class="exit"]
stmt2 [height=2]
stmt3 [height=2]
entry -> stmt1
stmt1 -> return1 [class="consequence"]
stmt1 -> stmt2 [class="alternative"]
stmt2 -> return2 [class="consequence"]
stmt2 -> stmt3 [class="alternative"]
stmt3 -> stmt4_1
stmt4_1 -> stmt4_2
stmt4_2 -> return4_2 [class="consequence"]
stmt4_2 -> stmt4_4 [class="alternative"]
stmt4_4 -> stmt4_1 [dir="back"]
stmt4_4 [height=3]
\`\`\`
`
    );
    expect(result).toMatchSnapshot();
  });

  test("Render CFG with clusters", ()=>{
    const result  = md.render(
      `
\`\`\`dot-cfg
subgraph cluster_tryComplex {
    class=tryComplex
    subgraph cluster_try {
        class=try
        TRY [label=try]
    }
    subgraph cluster_except {
        class=except
        EXCEPT [label="except/catch"]
    }
    subgraph cluster_else {
        ELSE [label=else]
    }
    subgraph cluster_finally {
        class=finally
        FINALLY [label=finally]
    }
}
ENTRY -> TRY
TRY -> EXCEPT [class=exception]
TRY -> ELSE
ELSE -> FINALLY
EXCEPT -> FINALLY
FINALLY -> EXIT
\`\`\`
`
    );
    expect(result).toMatchSnapshot();
  });
})