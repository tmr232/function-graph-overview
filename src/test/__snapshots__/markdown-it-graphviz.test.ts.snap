// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Render DOT Samples Render DOT 1`] = `
"<p>Here's a DOT snippet!</p>
<pre><code class="language-dot">digraph G {

  subgraph cluster_0 {
    style=filled;
    color=lightgrey;
    node [style=filled,color=white];
    a0 -&gt; a1 -&gt; a2 -&gt; a3;
    label = &quot;process #1&quot;;
  }

  subgraph cluster_1 {
    node [style=filled];
    b0 -&gt; b1 -&gt; b2 -&gt; b3;
    label = &quot;process #2&quot;;
    color=blue
  }
  start -&gt; a0;
  start -&gt; b0;
  a1 -&gt; b3;
  b2 -&gt; a3;
  a3 -&gt; a0;
  a3 -&gt; end;
  b3 -&gt; end;

  start [shape=Mdiamond];
  end [shape=Msquare];
}
</code></pre>
"
`;

exports[`Render DOT-CFG Samples Render CFG 1`] = `
"<pre><code class="language-dot-cfg">entry [class=&quot;entry&quot;]
return1 [class=&quot;exit&quot;]
return2 [class=&quot;exit&quot;]
return4_2 [class=&quot;exit&quot;]
stmt2 [height=2]
stmt3 [height=2]
entry -&gt; stmt1
stmt1 -&gt; return1 [class=&quot;consequence&quot;]
stmt1 -&gt; stmt2 [class=&quot;alternative&quot;]
stmt2 -&gt; return2 [class=&quot;consequence&quot;]
stmt2 -&gt; stmt3 [class=&quot;alternative&quot;]
stmt3 -&gt; stmt4_1
stmt4_1 -&gt; stmt4_2
stmt4_2 -&gt; return4_2 [class=&quot;consequence&quot;]
stmt4_2 -&gt; stmt4_4 [class=&quot;alternative&quot;]
stmt4_4 -&gt; stmt4_1 [dir=&quot;back&quot;]
stmt4_4 [height=3]
</code></pre>
"
`;

exports[`Render DOT-CFG Samples Render CFG with clusters 1`] = `
"<pre><code class="language-dot-cfg">subgraph cluster_tryComplex {
    class=tryComplex
    subgraph cluster_try {
        class=try
        TRY [label=try]
    }
    subgraph cluster_except {
        class=except
        EXCEPT [label=&quot;except/catch&quot;]
    }
    subgraph cluster_else {
        ELSE [label=else]
    }
    subgraph cluster_finally {
        class=finally
        FINALLY [label=finally]
    }
}
ENTRY -&gt; TRY
TRY -&gt; EXCEPT [class=exception]
TRY -&gt; ELSE
ELSE -&gt; FINALLY
EXCEPT -&gt; FINALLY
FINALLY -&gt; EXIT
</code></pre>
"
`;
